### DISCUSSION TITLE
How to rotate the Phantom 3d Camera left and right for tank controls for existing player character

### DISCUSSION THREAD

**Discussion by VoltitanDev** [ramokz/phantom-camera#227] [Question / Help Wanted] [answered]
Hi, I'm a bit new to Godot 4.2 and was trying out the Phantom Camera 3D but I was wondering how would I apply the 3d rotation of the Phantom camera to the player's left and right input as well as have the player be the center of the aforementioned rotation?

**Comment by ramokz** [comment]
If you've got an image or video example then it might help figure out an ideal solution, but I'm guessing you are looking for some sort of orbit camera position, where pressing left / right rotates the camera around the player?

**Reply by VoltitanDev** [reply]
Yes, with the third person follow mode https://www.youtube.com/watch?v=Q_S_TXHzKY0
This is from an older version of my current project.

**Reply by ramokz** [reply]
That should be simple enough.
Instead of using mouse movement, like in the example scene inside the project, you can adjust the rotation of the `PCam3D` using key events.

E.g.

```gdscript
func _physics_process(delta: float) -> void:
  # Get the inputs that corresponds to the WASD / arrow keys from the project settings
  var input_dir: Vector2 = Input.get_vector(
    "move_left",
    "move_right",
    "move_up",
    "move_down"
  )

  # Store the current rotational value of the PCam3D
  var pcam_rotation_degrees: Vector3 = _player_pcam.get_third_person_rotation_degrees()

  # Update the rotational value of the y-axis (yaw) 
  pcam_rotation_degrees.y -= input_dir.x * 2 # Replace 2 with desired rotational speed

  # Apply the new rotation back to the PCam3D
  _player_pcam.set_third_person_rotation_degrees(pcam_rotation_degrees)
```

**Reply by VoltitanDev** [reply]
Your example helped resolve the turning issue but, I noticed that the camera's following is inconsistent for instance, when my player character lands a certain way after flying(For context, my game is a third-person super robot shooter), the camera looks at the player either from a different angle or in front of the player.

**Reply by ramokz** [reply]
My guess would be that the camera rotation is changed as you fly or when you land, where you would need to reset its position to the intended value. You can use a Lerp or Tween function to make the transition feel seamless.

**Reply by VoltitanDev** [reply]
This is the video example of the issue so far.

https://github.com/ramokz/phantom-camera/assets/162152507/5c722a87-4da4-47a7-b104-efcbd5a18b70

edit: So far I resolved the issue when he flies by manipulating the x rotation from the phantom camera looking at the player but I'm still getting the issue where the camera looks at the player differently depending on where he lands.

**Reply by ramokz** [reply]
Nothing immediately obvious springs to mind when seeing that, but a couple of things to check:
- Has the `follow target`'s position changed? I.e. is it possible that the exact spot it is tracking isn't in the position it ought to be?
- Has the rotation of the `PCam3D`changed in any way, or is it still (0, 0, 0)?

**Reply by VoltitanDev** [reply]
I believe it's the former since the camera  does rotates when the player moves  though rotation of the camera has changed within the editor to have it face the player's back within a certain angle.

**Reply by ramokz** [reply]
Sorry for the delayed response.

I'm guessing it's a case of then applying a specific position and angle every time the character goes between the different areas (land and air). So that there is a hard-coded position and perspective being applied, so you control exactly how it should behave, rather than the system guessing what the intent should be.

**Reply by VoltitanDev** [reply]
Hi, sorry for the late reply I've tried to get the camera to rotate to the mouse similarly to the Third person example but the mouse isn't as responsive as I hoped.  For context the I set up the both the main and phantom camera as  Export Variables.
[Player_script.zip](https://github.com/ramokz/phantom-camera/files/14795044/Player_script.zip)

**Reply by ramokz** [reply]
It's a bit difficult to debug and validate from a single script alone as many other factors could play into it like physics collisions, other scripts etc.
If you're able to share a minimal reproducible project (MRP) where that script is being used, don't worry about the environment or model, then that would make it a lot easier.

**Comment by VoltitanDev** [comment]
What's a MRP?

**Reply by ramokz** [reply]
Minimal Reproducible Project

**Comment by VoltitanDev** [comment]
Voltitan_vol3_MRP.zip
<https://drive.google.com/file/d/1JGF4NAsWYDGy2yepQTeYDUI0X6hGUqXg/view?usp=drive_web>

**Comment by VoltitanDev** [comment]
Voltitan_vol3_mrp.zip
<https://drive.google.com/file/d/1zP8B81vbdqCDFqQktHtIBcmadSnqih6W/view?usp=drive_web>
Sorry the other folder was having some dependency issues here's another
folder. The Minimal Scene is the Test2 scene from the Level folder and can
be played by running the project

**Reply by ramokz** [reply]
Thanks for sharing this!

It seems to be account restricted though? At least, can't seem to open the link without a Google account.

**Comment by VoltitanDev** [comment]
Sorry about that, fixed the permissions
 Voltitan_vol3_mrp.7z
<https://drive.google.com/file/d/1eWan7JqNpN93MSmM6cDOzSlPWndIOEeY/view?usp=drive_web>

**Reply by ramokz** [reply]
That link works!
So opened up the project and the scene you mentioned, but I am a bit lost as to how it works. It looks and behaves differently than the video you shared earlier and am not sure where to start?

https://github.com/ramokz/phantom-camera/assets/5159399/efb1b7a3-3849-41f2-bc01-5ebf5295cdd6

**Reply by VoltitanDev** [reply]
It turns out the camera responds to the mouse input when I switched the unhandled input function the _input function event instead.

**Reply by ramokz** [reply]
If `_input()` works but `unhandled_input()` does not, then that would suggest that something else is capturing the focus / input.
The reason `_unhandled_input()` is generally what should be used for gameplay is because if you have, say, a UI menu popping up, then you would ideally want to prevent the gameplay controls from responding to, in this case, mouse movement.

**Reply by VoltitanDev** [reply]
The function that captured the input would be the get_input(delta) function from the player's script since that function handles the player's controls on land.

**Comment by VoltitanDev** [comment]
The level scene in question is the test2 scene within the Level folder. The player's behavior earlier was before I attempted to have the camera use the mouse inputs like the 3rd person controller example scene. I was having trouble having the phantom camera respond to the mouse inputs.

**Reply by ramokz** [reply]
Are you still having issues then? Since I'm not entirely sure what I should be looking for in the project files.

**Reply by VoltitanDev** [reply]
The player within the Test2 scenes is the one with the problem with the unhandled inputs as well as the one using the third person phantom camera. So far the camera still responds to the _input function rather than the unhandled input.

**Reply by ramokz** [reply]
Sorry, got distracted with the recent release and getting fixes in.
I had a look again, but it's a bit difficult to figure out what might be causing it, as a lot of things are happening in the project / scripts.

Would suggest creating a new scene / project that only has the bare minimum setup for the controls and camera. That way, you can start isolating the possible variables that cause the issue. To the previous comment, it's highly likely that something else is overtaking the input controls, but it's difficult to spot what that might be with so much going on.

That's also the intent behind MRP (minimal reproducible project), where only the essentials are present to avoid conflicting logic that larger projects contain. It is also helpful for others who are unfamiliar with a project by not having to debug larger codebases that might be interfering, and instead just focus on a specific issue.

**Comment by VoltitanDev** [comment]
The Inputs themselves are actually handled by the get_input function that's
called through the physics_process function of the player(VolTitan) script.
I think that might also be a factor for the issue with the unhandled input
not responding to the mouse.

**Comment by VoltitanDev** [comment]
Here's new MRP that isolates the clutter.
 Voltitan_vol3_mrp2.zip
<https://drive.google.com/file/d/1SjfvGAm5_tZtLFeLhU5ngIkC4LZ2xqBp/view?usp=drive_web>

**Reply by ramokz** [reply]
Am not sure which file to open? Opening the project triggers an error for a Beta scene, and the only scenes that are open by default are the addon's Third-Person example scene and `MainMenu`.

Think the issue you're facing here is more of a general Godot input conflict rather than something specific to the addon. I, personally, prefer to help out if it's related to Phantom Camera specifically — in particular when it's on a more complicated project such as yours. As mentioned earlier, if the `_unhandled_input` doesn't catch a particular input, then that means something else in the project does. To me, that feels more appropriate for a general Godot troubleshooting forum than here, as the goal seems to be about recognising input rather than a certain part of the addon isn't working. The addon is meant to work off the back of recognising the input, but the input recognition is a general Godot project-specific challenge.

For some more context about MRP, the intent behind a minimal reproducible project (MRP), is that it only contains a single scene with just the content to get the basic behaviour working and to highlight what isn't working. Meaning, no additional script, no assets or fancy stuff in general. It's so that there's no conflicting logic that's interfering, but also so that the issue reporter can potentially self-identify the issue by isolating the problem to as few lines of code as possible. E.g. an addon specific function doesn't work correctly or an error being triggered when toggling something on or off.

**Reply by VoltitanDev** [reply]
The beta scene is the mrp scene with just the player and the the phantom cameras and main camera. But I was wondering for an already established player character do I need to use the Unhandled_Input to use the phantom camera? The player's inputs are handled by the get_input(delta)  function. The camera still works when using the _input(event) instead of unhandled_input

**Reply by ramokz** [reply, marked as answer]
Sure, but the Beta scene is still throwing an error when trying to open it:
![image](https://github.com/ramokz/phantom-camera/assets/5159399/7aae666d-b4a2-4189-9807-b5ef68870a8a)

To answer your question, you ought to be using `_unhandled_input` for gameplay related inputs. Like its name implies, it only triggers when something else doesn't catch inputs. E.g. if you have a pause menu open, you probably don't want the playable character, and PCams, to be controllable at that point. Godot covers that in more detail [on this page](https://docs.godotengine.org/en/latest/tutorials/inputs/inputevent.html#how-does-it-work).

My point in the previous post is that the addon itself isn't the underlying issue, but rather that there's a conflicting input in the other scripts in the project/scene(s). So the help request that I'm hearing isn't specific to the project, but rather an input management issue for a particular project. Something, I think, is more appropriate in a general Godot forum, as technically anyone would be able to help

**Reply by VoltitanDev** [reply]
Thanks for the suggestion, and the Phantom camera responds with the unhandled_input within the player's controller it turns out its the control nodes within the level's scene that are interfering with the mouse's input.
